### PR TITLE
Remove type field from DocWriteRequest and associated Response objects

### PR DESCRIPTION
**Pull request by romseygeek**
This commit removes the `type` field from index, update and delete requests, and their
associated responses.

Relates to #41059

### CODE REVIEW

**Comment by romseygeek** [comment]
Note that this PR does not change the methods on `Client`, as there are over 1,000 instances of `prepareIndex(index, type, id) in the tests and the code changes here are already enormous.  I'll change those methods in a followup.

The rest high level client also directly uses these classes, which makes this a breaking change.  One step to consider would be to clone them into the RHLC module, as we have already done with many other request/response classes, and then mark the typed constructors/setters as deprecated.

**Review by jtibshirani — COMMENTED** [review]
This looks like a good direction to me! It looks like tests are failing, so I'll hold off on a full review.

> The rest high level client also directly uses these classes, which makes this a breaking change. One step to consider would be to clone them into the RHLC module, as we have already done with many other request/response classes, and then mark the typed constructors/setters as deprecated.

I'm not sure why we would copy these request/ response classes as part of types removal work? In 7.x we copied other requests/ responses like `CreateIndexRequest` because for these classes, it was very difficult to support both typed and typeless calls at once. For the classes in this PR, we didn't run into the same problems -- and now in 8.x we can simply remove the deprecated typed constructors and setters like `IndexRequest#type`.

**Comment by romseygeek** [comment]
> and now in 8.x we can simply remove the deprecated typed constructors and setters like IndexRequest#type.

You're right, I got confused by things not being deprecated on `DocWriteRequest`, but everything is indeed marked as deprecated in the concrete child classes, so no refactoring necessary.

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java [hunk 1]
```diff
@@ -684,7 +684,8 @@ private static void deleteAllILMPolicies() throws IOException {
             Response response = adminClient().performRequest(new Request("GET", "/_ilm/policy"));
             policies = entityAsMap(response);
         } catch (ResponseException e) {
-            if (RestStatus.METHOD_NOT_ALLOWED.getStatus() == e.getResponse().getStatusLine().getStatusCode()) {
+            if (RestStatus.METHOD_NOT_ALLOWED.getStatus() == e.getResponse().getStatusLine().getStatusCode() ||
```
Comment (jtibshirani) [inline test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java]: Looks like some leftover code to work around test failures?

File: x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardChangesActionTests.java [hunk 2]
```diff
@@ -147,7 +147,7 @@ public void testGetOperationsExceedByteLimit() throws Exception {
         final IndexShard indexShard = indexService.getShard(0);
         final Translog.Operation[] operations = ShardChangesAction.getOperations(indexShard, indexShard.getLastKnownGlobalCheckpoint(),
             0, 12, indexShard.getHistoryUUID(), new ByteSizeValue(256, ByteSizeUnit.BYTES));
-        assertThat(operations.length, equalTo(12));
+        assertThat(operations.length, equalTo(11));
```
Comment (jtibshirani) [inline x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardChangesActionTests.java]: For my context, why did this number change?

**Review by jtibshirani — COMMENTED** [review]
What a big effort! This looks good to me, I just left a couple questions/ comments.

File: docs/reference/mapping/removal_of_types.asciidoc [hunk 3]
```diff
@@ -532,7 +532,6 @@ PUT index/_doc/1
 {
   "_index": "index",
   "_id": "1",
-  "_type": "_doc",
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply docs/reference/mapping/removal_of_types.asciidoc]: Possibly, although I imagine we will remove this page entirely, and instead have something in the upgrade notes - @colings86 @jpountz do you have an opinion on what to do here?

File: server/src/main/java/org/elasticsearch/action/delete/DeleteRequestBuilder.java [hunk 4]
```diff
@@ -41,9 +41,10 @@ public DeleteRequestBuilder(ElasticsearchClient client, DeleteAction action, @Nu
 
     /**
      * Sets the type of the document to delete.
+     * @deprecated types are being removed
      */
+    @Deprecated
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/action/delete/DeleteRequestBuilder.java]: ++

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java (hunk 1 above)
Comment (romseygeek) [reply test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java]: So this was an interesting one; all the REST tests started failing with 400 errors, and it took me a while to track down what was happening.  Previously, if SLM or ILM were not enabled, a call to `GET _ilm/policy` would be interpreted as an index matching `{index}/{type}`, which only has handlers registered for `PUT` and `POST`, and so would return a `406 Method not allowed` error.  Now, the `{index}/{type}` handler has gone away, and so we just get a `400 Bad Request` instead.  I don't think we need to check for `406` errors any more, so I'll try removing that line.

File: x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardChangesActionTests.java (hunk 2 above)
Comment (romseygeek) [reply x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardChangesActionTests.java]: The test is checking what happens when documents stored in the translog take up more than a certain amount of space, in this case 256 bytes.  Previously, the documents had a type of `doc`, but now they have a type of `_doc` (as a default - a future PR will remove type entirely from the doc stored in the translog), which is one byte longer, so we fit one fewer document in the same number of bytes.

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java (hunk 1 above)
Comment (romseygeek) [reply test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java]: Looks like we still need to check for both due to BWC checks

Comment (jtibshirani) [reply test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java]: Got it, this change makes sense to me.

File: x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardChangesActionTests.java (hunk 2 above)
Comment (jtibshirani) [reply x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardChangesActionTests.java]: Thanks for the explanation.

**Review by jtibshirani — APPROVED** [review]
This looks good to me, apart from the question around `removal_of_types.asciidoc`.

**Comment by romseygeek** [comment]
Thanks @jtibshirani - I added an item to the meta issue to discuss what to do about the `removal_of_types` documentation.

**Pull request merged**